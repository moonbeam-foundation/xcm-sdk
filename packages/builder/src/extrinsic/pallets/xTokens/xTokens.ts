import { ExtrinsicConfig } from '../../../types/substrate/ExtrinsicConfig';
import {
  type ExtrinsicConfigBuilder,
  XcmVersion,
} from '../../ExtrinsicBuilder.interfaces';
import {
  getExtrinsicArgumentVersion,
  normalizeConcrete,
  normalizeX1,
} from '../../ExtrinsicBuilder.utils';
import { getDestination, getWeight } from './xTokens.utils';

const pallet = 'xTokens';

export function xTokens() {
  return {
    transfer: (): ExtrinsicConfigBuilder => ({
      build: ({ destinationAddress, asset, destination }) =>
        new ExtrinsicConfig({
          module: pallet,
          func: 'transfer',
          getArgs: (func) => {
            const destIndex = 2;
            const version = getExtrinsicArgumentVersion(func, destIndex);

            return [
              asset.getAssetId(),
              asset.amount,
              getDestination(version, destinationAddress, destination),
              getWeight(),
            ];
          },
        }),
    }),
    transferMultiAsset: (originParachainId: number) => {
      const funcName = 'transferMultiasset';
      const destIndex = 1;

      return {
        here: (): ExtrinsicConfigBuilder => ({
          build: ({ destinationAddress: address, asset, destination }) =>
            new ExtrinsicConfig({
              module: pallet,
              func: funcName,
              getArgs: (func) => {
                const version = getExtrinsicArgumentVersion(func, destIndex);

                return [
                  {
                    [version]: {
                      id: normalizeConcrete(version, {
                        parents: 0,
                        interior: 'Here',
                      }),
                      fun: {
                        Fungible: asset.amount,
                      },
                    },
                  },
                  getDestination(version, address, destination),
                  getWeight(),
                ];
              },
            }),
        }),
        X1: (): ExtrinsicConfigBuilder => ({
          build: ({ destinationAddress: address, asset, destination }) =>
            new ExtrinsicConfig({
              module: pallet,
              func: funcName,
              getArgs: (func) => {
                const version = getExtrinsicArgumentVersion(func, destIndex);

                return [
                  {
                    [version]: {
                      id: normalizeConcrete(
                        version,
                        normalizeX1(version, {
                          parents: 1,
                          interior: {
                            X1: {
                              Parachain: originParachainId,
                            },
                          },
                        }),
                      ),
                      fun: {
                        Fungible: asset.amount,
                      },
                    },
                  },
                  getDestination(version, address, destination),
                  getWeight(),
                ];
              },
            }),
        }),
        X2: (): ExtrinsicConfigBuilder => ({
          build: ({ destinationAddress: address, asset, destination }) =>
            new ExtrinsicConfig({
              module: pallet,
              func: funcName,
              getArgs: (func) => {
                const version = getExtrinsicArgumentVersion(func, destIndex);

                return [
                  {
                    [version]: {
                      id: normalizeConcrete(version, {
                        parents: 1,
                        interior: {
                          X2: [
                            {
                              Parachain: originParachainId,
                            },
                            {
                              GeneralKey: asset.getAssetId(),
                            },
                          ],
                        },
                      }),
                      fun: {
                        Fungible: asset.amount,
                      },
                    },
                  },
                  getDestination(version, address, destination),
                  getWeight(),
                ];
              },
            }),
        }),
      };
    },
    transferMultiCurrencies: (): ExtrinsicConfigBuilder => ({
      build: ({ destinationAddress: address, asset, destination, fee }) =>
        new ExtrinsicConfig({
          module: pallet,
          func: 'transferMulticurrencies',
          getArgs: () => [
            [
              [asset.getAssetId(), asset.amount],
              [fee.getAssetId(), fee.amount],
            ],
            1,
            getDestination(XcmVersion.v3, address, destination),
            getWeight(),
          ],
        }),
    }),
  };
}
