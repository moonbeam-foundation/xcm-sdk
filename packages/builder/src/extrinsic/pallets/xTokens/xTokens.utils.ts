import type { AnyParachain } from '@moonbeam-network/xcm-types';
import type { XcmVersion } from '../../ExtrinsicBuilder.interfaces';
import { getExtrinsicAccount } from '../../ExtrinsicBuilder.utils';
import type { XTokensWeightLimit } from './xTokens.interfaces';

/**
 * It is safer to always use unlimited
 */
export function getWeight(): XTokensWeightLimit {
  return 'Unlimited';
}

export function getDestination(
  version: XcmVersion,
  address: string,
  destination: AnyParachain,
) {
  return {
    [version]: {
      parents: 1,
      interior: {
        X2: [
          {
            Parachain: destination.parachainId,
          },
          getExtrinsicAccount(address),
        ],
      },
    },
  };
}
