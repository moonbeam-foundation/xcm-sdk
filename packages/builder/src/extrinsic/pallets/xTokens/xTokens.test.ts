import { describe, expect, it } from 'vitest';

import {
  buildParachainParamsMock,
  buildParamsMock,
} from '../../../../fixtures';
import { xTokens } from './xTokens';

describe('xTokens', () => {
  describe('transfer', () => {
    const extrinsic = xTokens().transfer().build(buildParamsMock);

    it('should be correct config', () => {
      expect(extrinsic).toMatchSnapshot();
    });

    it('should get correct arguments', () => {
      expect(extrinsic.getArgs()).toMatchSnapshot();
    });
  });

  describe('transferMultiAsset', () => {
    describe('x1', () => {
      const extrinsic = xTokens()
        .transferMultiAsset(100)
        .X1()
        .build(buildParamsMock);

      it('should be correct config', () => {
        expect(extrinsic).toMatchSnapshot();
      });

      it('should get correct arguments', () => {
        expect(extrinsic.getArgs()).toMatchSnapshot();
      });
    });
    describe('x2', () => {
      const extrinsic = xTokens()
        .transferMultiAsset(100)
        .X2()
        .build(buildParamsMock);

      it('should be correct config', () => {
        expect(extrinsic).toMatchSnapshot();
      });

      it('should get correct arguments', () => {
        expect(extrinsic.getArgs()).toMatchSnapshot();
      });
    });
    describe('parachain', () => {
      const extrinsic = xTokens()
        .transferMultiAsset(100)
        .X2()
        .build(buildParachainParamsMock);

      it('should be correct config', () => {
        expect(extrinsic).toMatchSnapshot();
      });

      it('should get correct arguments', () => {
        expect(extrinsic.getArgs()).toMatchSnapshot();
      });
    });
  });

  describe('transferMultiCurrencies', () => {
    describe('x2', () => {
      const extrinsic = xTokens()
        .transferMultiCurrencies()
        .build(buildParamsMock);

      it('should be correct config', () => {
        expect(extrinsic).toMatchSnapshot();
      });

      it('should get correct arguments', () => {
        expect(extrinsic.getArgs()).toMatchSnapshot();
      });
    });

    describe('parachain', () => {
      const extrinsic = xTokens()
        .transferMultiCurrencies()
        .build(buildParachainParamsMock);

      it('should be correct config', () => {
        expect(extrinsic).toMatchSnapshot();
      });

      it('should get correct arguments', () => {
        expect(extrinsic.getArgs()).toMatchSnapshot();
      });
    });
  });
});
