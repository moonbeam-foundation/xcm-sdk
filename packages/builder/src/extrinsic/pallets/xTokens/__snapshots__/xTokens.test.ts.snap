// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`xTokens > transfer > should be correct config 1`] = `
ExtrinsicConfig {
  "func": "transfer",
  "getArgs": [Function],
  "module": "xTokens",
}
`;

exports[`xTokens > transfer > should get correct arguments 1`] = `
[
  "USDT",
  99000000000n,
  {
    "V1": {
      "interior": {
        "X2": [
          {
            "Parachain": 1000,
          },
          {
            "AccountKey20": {
              "key": "0xeF46c7649270C912704fB09B75097f6E32208b85",
            },
          },
        ],
      },
      "parents": 1,
    },
  },
  "Unlimited",
]
`;

exports[`xTokens > transferMultiAsset > parachain > should be correct config 1`] = `
ExtrinsicConfig {
  "func": "transferMultiasset",
  "getArgs": [Function],
  "module": "xTokens",
}
`;

exports[`xTokens > transferMultiAsset > parachain > should get correct arguments 1`] = `
[
  {
    "V1": {
      "fun": {
        "Fungible": 99000000000n,
      },
      "id": {
        "Concrete": {
          "interior": {
            "X2": [
              {
                "Parachain": 100,
              },
              {
                "GeneralKey": "USDT",
              },
            ],
          },
          "parents": 1,
        },
      },
    },
  },
  {
    "V1": {
      "interior": {
        "X2": [
          {
            "Parachain": 2032,
          },
          {
            "AccountId32": {
              "id": "0x18e6c7ad45fbab18778710642072fdd057470b76d4d568b44dea6accd5b0c423",
              "network": null,
            },
          },
        ],
      },
      "parents": 1,
    },
  },
  "Unlimited",
]
`;

exports[`xTokens > transferMultiAsset > x1 > should be correct config 1`] = `
ExtrinsicConfig {
  "func": "transferMultiasset",
  "getArgs": [Function],
  "module": "xTokens",
}
`;

exports[`xTokens > transferMultiAsset > x1 > should get correct arguments 1`] = `
[
  {
    "V1": {
      "fun": {
        "Fungible": 99000000000n,
      },
      "id": {
        "Concrete": {
          "interior": {
            "X1": {
              "Parachain": 100,
            },
          },
          "parents": 1,
        },
      },
    },
  },
  {
    "V1": {
      "interior": {
        "X2": [
          {
            "Parachain": 1000,
          },
          {
            "AccountKey20": {
              "key": "0xeF46c7649270C912704fB09B75097f6E32208b85",
            },
          },
        ],
      },
      "parents": 1,
    },
  },
  "Unlimited",
]
`;

exports[`xTokens > transferMultiAsset > x2 > should be correct config 1`] = `
ExtrinsicConfig {
  "func": "transferMultiasset",
  "getArgs": [Function],
  "module": "xTokens",
}
`;

exports[`xTokens > transferMultiAsset > x2 > should get correct arguments 1`] = `
[
  {
    "V1": {
      "fun": {
        "Fungible": 99000000000n,
      },
      "id": {
        "Concrete": {
          "interior": {
            "X2": [
              {
                "Parachain": 100,
              },
              {
                "GeneralKey": "USDT",
              },
            ],
          },
          "parents": 1,
        },
      },
    },
  },
  {
    "V1": {
      "interior": {
        "X2": [
          {
            "Parachain": 1000,
          },
          {
            "AccountKey20": {
              "key": "0xeF46c7649270C912704fB09B75097f6E32208b85",
            },
          },
        ],
      },
      "parents": 1,
    },
  },
  "Unlimited",
]
`;

exports[`xTokens > transferMultiCurrencies > parachain > should be correct config 1`] = `
ExtrinsicConfig {
  "func": "transferMulticurrencies",
  "getArgs": [Function],
  "module": "xTokens",
}
`;

exports[`xTokens > transferMultiCurrencies > parachain > should get correct arguments 1`] = `
[
  [
    [
      "USDT",
      99000000000n,
    ],
    [
      "RMRK",
      5000000000n,
    ],
  ],
  1,
  {
    "V3": {
      "interior": {
        "X2": [
          {
            "Parachain": 2032,
          },
          {
            "AccountId32": {
              "id": "0x18e6c7ad45fbab18778710642072fdd057470b76d4d568b44dea6accd5b0c423",
              "network": null,
            },
          },
        ],
      },
      "parents": 1,
    },
  },
  "Unlimited",
]
`;

exports[`xTokens > transferMultiCurrencies > x2 > should be correct config 1`] = `
ExtrinsicConfig {
  "func": "transferMulticurrencies",
  "getArgs": [Function],
  "module": "xTokens",
}
`;

exports[`xTokens > transferMultiCurrencies > x2 > should get correct arguments 1`] = `
[
  [
    [
      "USDT",
      99000000000n,
    ],
    [
      "RMRK",
      5000000000n,
    ],
  ],
  1,
  {
    "V3": {
      "interior": {
        "X2": [
          {
            "Parachain": 1000,
          },
          {
            "AccountKey20": {
              "key": "0xeF46c7649270C912704fB09B75097f6E32208b85",
            },
          },
        ],
      },
      "parents": 1,
    },
  },
  "Unlimited",
]
`;
